### PR TITLE
fix #5131, #5130: add email/guest login, simulation cog, help menu to navbar

### LINKED ISSUE #5131: 20221209 Full navbar with cog, sharing link, etc.
Finish up the header of the React app to include navigation. We can clean up the interface to the server if need be.

### PR DESCRIPTION
**Pull request by garsuga**
Needs to be tested, especially with login edge cases.

### CODE REVIEW

File: react/src/component/login/login.tsx [hunk 1]
```diff
@@ -0,0 +1,217 @@
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import * as Icon from "@fortawesome/free-solid-svg-icons";
+import React, { useContext, useEffect, useState } from "react"
+import { Button, Col, Container, Dropdown, Form, Image, Modal, Nav, Row } from "react-bootstrap"
+import { Navigate, Route, Routes, useNavigate, useParams } from "react-router"
+import { AppWrapper, AuthMethod, CAppName, CLoginStatus, CSchema } from "../../data/appwrapper"
```
Comment (robnagler) [inline react/src/component/login/login.tsx:6]: Is `C` a common react prefix? What does it mean? Component? Constant?

File: react/src/component/login/login.tsx [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+type LoginConfirmParams = {
+    token: string,
+    method: AuthMethod,
+    needCompleteRegistration: "0" | "1"
+}
+
+export const LoginConfirm = (props) => {
+    let { token, method, needCompleteRegistration } = useParams<LoginConfirmParams>();
+
+    switch(method) {
```
Comment (robnagler) [inline react/src/component/login/login.tsx:124]: Ditto on formatting.

File: react/src/component/login/login.tsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    }
+}
+
+
+
+export const LoginNeedCompleteRegistration = (props) => {
+    let appName = useContext(CAppName);
+    let navigate = useNavigate();
+    let onCompleteForm = (data: {[key: string]: any}) => {
+        fetch(`/auth-complete-registration`, {
```
Comment (robnagler) [inline react/src/component/login/login.tsx:138]: I would like to see all fetches modularized so that we can swap them out. The sends should be independent of the receives, that is, you send a message and the receive happens based on an event registration. This allows the app to be fully asynchronous. Promises are quasi-async and do not allow the app to manage flow of control. You cannot ignore a promise, say, if the app state changes. You have to write special code to handle each promise.

The implementation of this type of event registration can be simple at first, e.g. based on API names.

File: react/src/component/login/login.tsx [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    )
+}
+
+export const LoginExtraInfoForm = (props: { onComplete: ({displayName}) => void }) => {
+    let { onComplete } = props;
+    let [displayName, updateDisplayName] = useState<string>("");
+    return (
+        <>
+            <p>Please enter your full name to complete your Sirepo registration.</p>
+            <Form.Group as={Row} className="mb-3" controlId="formPlaintextEmail">
```
Comment (robnagler) [inline react/src/component/login/login.tsx:157]: I'm concerned about the proliferation of these constants which are implicitly coupled. mb-3 is implicit and so is formPlaintextEmail. These sh

File: react/src/component/login/login.tsx [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        <LoginExtraInfoForm onComplete={(data) => onCompleteForm(data)}/>
+    )
+}
+
+export const LoginExtraInfoForm = (props: { onComplete: ({displayName}) => void }) => {
+    let { onComplete } = props;
+    let [displayName, updateDisplayName] = useState<string>("");
+    return (
+        <>
+            <p>Please enter your full name to complete your Sirepo registration.</p>
```
Comment (robnagler) [inline react/src/component/login/login.tsx:156]: Get this from the react schema. It should not be hard coded.

File: react/src/component/login/login.tsx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    let appName = useContext(CAppName);
+
+    let getLoginComponent = (method: string): JSX.Element => {
+        switch(method) {
+            case "email":
+                return <LoginWithEmail/>
+            case "guest":
+                return <Navigate to={new AppWrapper(appName).getAppRootLink()}/>
+            default:
+                throw new Error(`could not handle login method=${method}`)
```
Comment (robnagler) [inline react/src/component/login/login.tsx:181]: "could not handle" > "unknown". Try to keep error messages tokenizable and short. You could even code them as unknown-login-method, which can then be parsed on the server side.

File: react/src/component/login/login.tsx [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        </Container>
+    )
+}
+
+export function LogoutRoot(props) {
+    let navigate = useNavigate();
+    let appName = useContext(CAppName);
+
+    useEffect(() => {
+        fetch(`/auth-logout/${appName}`).then(() => navigate(new AppWrapper(appName).getAppRootLink()));
```
Comment (robnagler) [inline react/src/component/login/login.tsx:212]: URLs need to be managed centrally.

File: react/src/component/login/login.tsx [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    })
+    
+
+    return <>Singing out...</>
```
Comment (robnagler) [inline react/src/component/login/login.tsx:216]: One of the big reasons for having a string table is that it can be proofread and spell checked easily.

File: react/src/component/reusable/navbar.scss [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    flex-flow: row nowrap;
+    width: 100%;
+    padding-left: 10px;
+    padding-right: 10px;
+    align-items: center;
+}
+
+.sr-navbar {
+    padding: 0 !important;
+    height: 60px;
```
Comment (robnagler) [inline react/src/component/reusable/navbar.scss:20]: Is this number completely relative or related to any other number?

File: react/src/component/reusable/navbar.scss [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+.sr-navbar-simulations-button {
+    text-decoration: underline;
+}
+
+.sr-navbar-container {
+    display: flex;
+    flex-flow: row nowrap;
+    width: 100%;
+    padding-left: 10px;
```
Comment (robnagler) [inline react/src/component/reusable/navbar.scss:13]: This should be a constant:

```txt
./src/component/simbrowser.scss:    margin-bottom: 10px;
./src/component/simbrowser.scss:    margin: 10px;
./src/component/login/login.scss:    padding: 10px;
./src/component/reusable/navbar.scss:    padding-left: 10px;
./src/component/reusable/navbar.scss:    padding-right: 10px;
```

File: react/src/component/simulation.tsx [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+            fetch(`/simulation/${appName}/${simulationId}/0/source`).then(async (resp) => {
+                let simulationInfo = await resp.json();
+                let models = simulationInfo['models'] as ModelState[];
+
+                for(let [modelName, model] of Object.entries(models)) {
+                    modelsWrapper.updateModel(modelName, model);
+                }
+
+                resolve({...simulationInfo, simulationId});
+                updateHasInit(true);
```
Comment (robnagler) [inline react/src/component/simulation.tsx:67]: There has to be a better way to manage state than cascading booleans. This relates to the state machine problem for messages.

**Review by robnagler — COMMENTED** [review]
Here are the main points:
* fully async message passing, not fetch/resolve
* state management needs to be understood globally (remove isolated booleans)
* centralize constants including strings, urls, and css
